File: java/src/main/java/sammancoaching/DiagramPhysicalPrinter.java
package sammancoaching;

import java.io.IOException;
import java.util.concurrent.locks.ReentrantLock;
import java.util.logging.Logger;
import java.io.File;

/**
 * This is a class you'd like to get under test so you can change it safely.
 */
public class DiagramPhysicalPrinter {
    private final PhysicalPrinter physicalPrinter;
    private final PrintQueue printQueue;
    private static final Logger logger = Logger.getLogger(DiagramPhysicalPrinter.class.getName());
    private static final ReentrantLock mutex = new ReentrantLock();

    public DiagramPhysicalPrinter(PhysicalPrinter physicalPrinter, PrintQueue printQueue) {
        this.physicalPrinter = physicalPrinter;
        this.printQueue = printQueue;
    }

    public DiagramPhysicalPrinter() {
        this.physicalPrinter = new PhysicalPrinter();
        this.printQueue = new PrintQueue(this.physicalPrinter);
    }

    public boolean doPrint(FlowchartDiagram diagram, DiagramMetadata info, String targetFilename) throws IOException {
        PrinterDriverFactory factory = PrinterDriverFactory.getInstance();
        DiagramPrintDriver printerDriver = factory.createDriverForPrint();
        printerDriver.setDiagram(diagram);

        PrintMetadata data = new PrintMetadata(info.fileType);
        boolean success = false;

        try {
            mutex.lock();

            if (!physicalPrinter.isAvailable()) {
                logger.info("Physical Printer Unavailable");
            } else if (physicalPrinter.getJobCount() < 0) {
                logger.info("Physical Printer Unavailable Due to Job Count Inconsistency");
            } else {
                // Print the diagram using the Physical Printer
                printQueue.add(data);
                String summaryInformation = diagram.getSummaryInformation();
                logger.info("Diagram Summary Information: " + summaryInformation);
                boolean isSummary = summaryInformation.length() > 10;

                if (physicalPrinter.startDocument(!isSummary, false, "DiagramPhysicalPrinter")) {
                    if (printerDriver.printTo(physicalPrinter)) {
                        logger.info("Physical Printer Successfully printed");
                        success = true;
                    }

                    physicalPrinter.endDocument();
                }
            }

            if (success) {
                // Save a backup of the printed document as PDF
                File file = new File(data.getFilename());
                if (file.exists()) {
                    logger.info("Saving backup of printed document as PDF to file " + targetFilename);
                    diagram.getFlowchartAsPdf().copyFile(data.getFilename(), targetFilename, true);
                }
            }
        } catch (Exception e) {
            logger.severe("Failed to print document: " + e.getMessage());
            success = false;
        } finally {
            mutex.unlock();
            printerDriver.releaseDiagram();
        }

        return success;
    }
}

File: cpp/src/diagram_physical_printer.cpp
#include "diagram_physical_printer.h"
#include "physical_printer.h"

#include <filesystem>
#include <iostream>
#include <mutex>

#include "printer_driver_factory.h"
#include "diagram_printer.h"

DiagramPhysicalPrinter::DiagramPhysicalPrinter(PhysicalPrinter* physicalPrinter,
                                               PrintQueue* printQueue)
    : _physicalPrinter(physicalPrinter)
    , _printQueue(printQueue)
{
}

DiagramPhysicalPrinter::DiagramPhysicalPrinter()
{
    _physicalPrinter = new PhysicalPrinter();
    _printQueue = new PrintQueue(_physicalPrinter);
}

DiagramPhysicalPrinter::~DiagramPhysicalPrinter()
{
    if (_physicalPrinter) {
        delete _physicalPrinter;
        _physicalPrinter = nullptr;
    }
    if (_printQueue) {
        delete _printQueue;
        _printQueue = nullptr;
    }
}

bool DiagramPhysicalPrinter::DoPrint(const PrintableDiagram& printableDiagram,
                                   const DiagramMetadata& info,
                                   const std::string& targetFilename)
{
    DiagramPrintDriver* printerDriver = PrinterDriverFactory::GetInstance().CreateDriverForPrint();
    printerDriver->SetDiagram(printableDiagram.GetDiagram());

    PrintMetadata data(info.getFileType());
    std::mutex _mutex;
    bool success = false;

    try {
        _mutex.lock();

        const auto& tonerLevels = _physicalPrinter->GetTonerLevels();

        if (!_physicalPrinter->IsAvailable() || !(tonerLevels.at(Toner::Black) > 0 &&
                                                  tonerLevels.at(Toner::Cyan) > 0 &&
                                                  tonerLevels.at(Toner::Magenta) > 0 &&
                                                  tonerLevels.at(Toner::Yellow) > 0)) {
            std::cout << "Physical Printer Unavailable" << std::endl;
            success = false;
        }
        else if (_physicalPrinter->GetJobCount() < 0) {
            std::cout << "Physical Printer Unavailable Due to Job Count Inconsistency" << std::endl;
            success = false;
        }
        else {
            // Print the diagram using the Physical Printer
            _printQueue->Add(data);
            std::string summaryInformation = printableDiagram.SummaryInformation();
            std::cout << "Diagram Summary Information: " << summaryInformation << std::endl;
            bool isSummary = summaryInformation.length() > 10;

            if (_physicalPrinter->StartDocument(!isSummary, false, "DiagramPhysicalPrinter")) {
                if (printerDriver->PrintTo(_physicalPrinter)) {
                    std::cout << "Physical Printer Successfully printed" << std::endl;
                    success = true;
                }
                _physicalPrinter->EndDocument();
            }
        }

        if (success) {
            // save a backup of the printed document as pdf
            if (std::filesystem::exists(data.GetFilename())) {
                std::cout << "Saving backup of printed document as PDF to file: " 
                         << targetFilename << std::endl;
                printableDiagram.PrintToFile(data.GetFilename(), targetFilename);
            }
        }
    }
    catch (const std::exception& e) {
        std::cerr << "Failed to print document: " << e.what() << std::endl;
        success = false;
    }

    _mutex.unlock();
    
    // Clean up the printer driver
    printerDriver->ReleaseDiagram();
    delete printerDriver;

    return success;
}